src/libcrun/net_device.c

```diff
@@ -0,0 +1,479 @@
+          if (err_data->error == 0)
+            continue;
+          return crun_make_error (err, -err_data->error, "netlink error while looking for interface `%s`", ifname);
+        }
+
+      if (nlh->nlmsg_type == RTM_NEWLINK)
+        {
+          struct ifinfomsg *ifi = NLMSG_DATA (nlh);
+          index = ifi->ifi_index;
+          return index;
+        }
+    }
+
+  if (index == 0)
+    return crun_make_error (err, 0, "could not find device `%s`", ifname);
+
+  return index;
+}
+
+static int
+wait_for_ack (int sock, uint32_t seq, char *recv_buffer, size_t recv_buffer_size, libcrun_error_t *err)
+{
+  struct nlmsghdr *nlh_recv = (struct nlmsghdr *) recv_buffer;
+  ssize_t len;
+
+  do
+    {
+      len = TEMP_FAILURE_RETRY (recv (sock, recv_buffer, recv_buffer_size, 0));
+      if (len < 0)
+        return crun_make_error (err, errno, "recv");
+  } while (nlh_recv->nlmsg_seq != seq);
+
+  if (! NLMSG_OK (nlh_recv, len))
+    return crun_make_error (err, 0, "received invalid packet");
+
+  if (nlh_recv->nlmsg_type == NLMSG_ERROR)
+    {
+      struct nlmsgerr *err_data = (struct nlmsgerr *) NLMSG_DATA (nlh_recv);
+      if (err_data->error == 0)
+        return 0;
+      return crun_make_error (err, -err_data->error, "netlink error");
+    }
+
+  return crun_make_error (err, 0, "internal error: received unknown netlink packet type");
+}
+
+static int
+get_ip_addresses (int sock, uint32_t ifindex, char **out_ips, char *buffer, size_t buffer_size, libcrun_error_t *err)
+{
+  struct nl_req *req = (struct nl_req *) buffer;
+  cleanup_free char *ips = NULL;
+  size_t ips_len = 0;
+  struct ip_addr *ip;
+  uint32_t seq;
+  ssize_t len;
+  int ret;
+
+  seq = reset_request (req, RTM_GETADDR, NLM_F_DUMP | NLM_F_REQUEST, sizeof (struct ifaddrmsg));
+  req->ifa.ifa_family = AF_UNSPEC;
+
+  ret = send_request (sock, req, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  while ((len = TEMP_FAILURE_RETRY (recv (sock, buffer, buffer_size, 0))) > 0)
+    {
+      struct ifaddrmsg *ifa;
+      struct nlmsghdr *nlh;
+      struct rtattr *rta;
+      int rta_len;
+
+      nlh = (struct nlmsghdr *) buffer;
+
+      while (NLMSG_OK (nlh, len))
+        {
+          if (nlh->nlmsg_seq != seq)
+            continue;
+
+          if (nlh->nlmsg_type == NLMSG_DONE)
+            {
+              *out_ips = ips;
+              ips = NULL;
+              return 0;
+            }
+
+          if (nlh->nlmsg_type == NLMSG_ERROR)
+            {
+              struct nlmsgerr *err_data = (struct nlmsgerr *) NLMSG_DATA (nlh);
+              if (err_data->error == 0)
+                continue;
+              return crun_make_error (err, -err_data->error, "netlink error reading ip addresses");
+            }
+
+          ifa = (struct ifaddrmsg *) NLMSG_DATA (nlh);
+          if (ifa->ifa_index != ifindex)
+            {
+              nlh = NLMSG_NEXT (nlh, len);
+              continue;
+            }
+
```

Comment:
Oh and I just fully read the spec

> The runtime MUST preserve existing network interface attributes, including all
permanent IP addresses (IFA_F_PERMANENT flag) of any family with global scope
(RT_SCOPE_UNIVERSE value) as defined in [`RFC 3549 Section 2.3.3.2`][rfc3549].
This ensures that only addresses intended for persistent, external communication
are transferred. 

You already apply IFA_F_PERMANENT to the address when you set them but this here doesn't seem to skip addresses without IFA_F_PERMANENT set? I guess we would not want to copy an ipv6 link local address.